### PR TITLE
handeling numbers in words for chinese

### PR DESCRIPTION
**Pull request by harjinder7**
## JIRA Ticket Number

JIRA TICKET: ML-3335

## Description of change
handeling numbers in words for chinese

### CODE REVIEW

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 1]
```diff
@@ -46,6 +45,8 @@ def __init__(self, entity_name='number', unit_type=None):
             '+': '加'
         }
 
+        self.power_of_10 = { 10 ** i for i in range(1,17)}
```
Comment (haptik-deployment) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:48]: E201 whitespace after '{'
E231 missing whitespace after ','

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 2]
```diff
@@ -149,4 +151,73 @@ def get_number_digit_by_digit(self, text=''):
...
+        result = ''
+        if not text:
+            return result
+
+        # extract the digits and scales from the text
+        digit_list = []
+
+        pwr_index_map = {}
+        for t in text:
+            digit_val = self.base_numbers_map_full.get(t)
```
Comment (haptik-deployment) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:163]: E711 comparison to None should be 'if cond is None:'

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 3]
```diff
@@ -149,4 +151,73 @@ def get_number_digit_by_digit(self, text=''):
...
+            else:
+                digit_list.append(digit_val)
+                if digit_val in self.power_of_10:
+                    pwr_index_map[digit_val] = len(digit_list) - 1
+        if len(digit_list) == 0:
+            return result
+
+        pwr_index_map[1] = len(digit_list)
+
+        # starting from highest power aggregate digit and scales
```
Comment (haptik-deployment) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:175]: E251 unexpected spaces around keyword / parameter equals
E203 whitespace before ':'

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 4]
```diff
@@ -149,4 +151,73 @@ def get_number_digit_by_digit(self, text=''):
...
+        pwr_index_list = sorted(pwr_index_map.items(), key= lambda kv : kv[0], reverse=True)
+
+        st = 0
+        final_val = 0
+        for pwr, indx in pwr_index_list:
```
Comment (haptik-deployment) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:180]: E201 whitespace after '['
E203 whitespace before ':'

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 5]
```diff
@@ -149,4 +151,73 @@ def get_number_digit_by_digit(self, text=''):
...
+            value = self.combine_digit_and_scale(digit_list[ st : indx])
+            st = indx + 1
```
Comment (haptik-deployment) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:182]: E711 comparison to None should be 'if cond is not None:'

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 6]
```diff
@@ -149,4 +151,73 @@ def get_number_digit_by_digit(self, text=''):
...
+        for i in range(start_index, len(num_list)):
+            x = num_list[i]
+            if x == 0:
+                zero_found = True
+                continue
+            if x in self.power_of_10:
+                digit_scaled_list[-1] = digit_scaled_list[-1] * x
+            else:
+                digit_scaled_list.append(x)
+        value = sum(digit_scaled_list)
```
Comment (haptik-deployment) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:220]: E201 whitespace after '('
E712 comparison to False should be 'if cond is False:' or 'if not cond:'

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 7]
```diff
@@ -149,4 +151,73 @@ def get_number_digit_by_digit(self, text=''):
...
```
Comment (haptik-deployment) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:220]: E712 comparison to False should be 'if cond is False:' or 'if not cond:'

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 8]
```diff
@@ -46,6 +45,8 @@ def __init__(self, entity_name='number', unit_type=None):
             '+': '加'
         }
 
+        self.power_of_10 = {10 ** i for i in range(1, 17)}
```
Comment (naseem-shaik) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:48]: Please perform multiplication over power operation to achieve the same.

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 9]
```diff
@@ -149,4 +151,73 @@ def get_number_digit_by_digit(self, text=''):
...
+        return : int
+        example :
+            [ 2, 1000, 5, 100, 3, 10, 8 ]
+            output : 2538
+        """
+        value = None
+        num_list = num_list or []
+
+        if len(num_list) == 0:
+            return value
```
Comment (naseem-shaik) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:202]: Is this case expected, if you are handling in the case of errors, please log an error.

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 10]
```diff
@@ -149,4 +151,73 @@ def get_number_digit_by_digit(self, text=''):
...
+
+        result = str(final_val)
+        return result
+
+    def combine_digit_and_scale(self, num_list: None):
+        """
+        params : list of numbers either digit or scale
+        return : int
+        example :
+            [ 2, 1000, 5, 100, 3, 10, 8 ]
```
Comment (naseem-shaik) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:195]: Please add this example too: 

2393576: 2 100 3 10 9 10000 3 1000 5 100 7 10

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 11]
```diff
@@ -149,4 +151,73 @@ def get_number_digit_by_digit(self, text=''):
 
     def get_number_with_digit_scaling(self, text=''):
```
Comment (naseem-shaik) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:152]: Add a description of what it does with the same example. 

2393576: sent in two batches [2 100 3 10 9], [ 3 1000 5 100 7 10]

**Review by naseem-shaik — APPROVED** [review]
LGTM, just a couple of descriptions to be added.

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 12]
```diff
@@ -149,4 +151,73 @@ def get_number_digit_by_digit(self, text=''):
...
+            value = self.combine_digit_and_scale(digit_list[st: indx])
+            st = indx + 1
+            if value is not None:
+                value = value * pwr
+                final_val += value
+
+        result = str(final_val)
+        return result
+
+    def combine_digit_and_scale(self, num_list: None):
```
Comment (naseem-shaik) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:190]: we are assigning the type to the default value.

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py (hunk 8 above)
Comment (harjinder7) [reply ner_v2/detectors/numeral/number/zh-TW/number_detection.py]: changes are done

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py (hunk 9 above)
Comment (harjinder7) [reply ner_v2/detectors/numeral/number/zh-TW/number_detection.py:235]: this is an expected case not an error.

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py (hunk 10 above)
Comment (harjinder7) [reply ner_v2/detectors/numeral/number/zh-TW/number_detection.py]: example is updated

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py (hunk 11 above)
Comment (harjinder7) [reply ner_v2/detectors/numeral/number/zh-TW/number_detection.py:156]: description updated

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py (hunk 12 above)
Comment (harjinder7) [reply ner_v2/detectors/numeral/number/zh-TW/number_detection.py]: this is also fixed

**Review by naseem-shaik — DISMISSED** [review]

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 13]
```diff
@@ -130,15 +135,27 @@ def _detect_number_from_text(self, number_list=None, original_list=None):
                     original_list.append(original_text)
         return number_list, original_list
 
+    def check_sign(self, text=''):
+        signs = {'-','+'}
```
Comment (haptik-deployment) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:139]: E231 missing whitespace after ','

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 14]
```diff
@@ -130,15 +135,27 @@ def _detect_number_from_text(self, number_list=None, original_list=None):
...
+        sign = None
+        new_text = text
+        for i in range(len(text)-1, -1, -1):
```
Comment (haptik-deployment) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:142]: E226 missing whitespace around arithmetic operator

File: ner_v2/detectors/numeral/number/zh-TW/number_detection.py [hunk 15]
```diff
@@ -130,15 +135,27 @@ def _detect_number_from_text(self, number_list=None, original_list=None):
...
+            if text[i] in signs:
+                sign = text[i]
+                new_text = text[i+1:]
```
Comment (haptik-deployment) [inline ner_v2/detectors/numeral/number/zh-TW/number_detection.py:145]: E226 missing whitespace around arithmetic operator

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Review by HiteshC6595 — APPROVED** [review]

**Pull request merged**